### ISSUE TITLE
[STORM-3593] Inconsistent library versions notice.

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#7375]
<p> <br/>
 <br/>
Hi. I have implemented a tool to detect library version inconsistencies. Your project have 10 inconsistent libraries and 2 false consistent libraries.<br/>
 <br/>
Take commons-io:commons-io for example, this library is declared as version 2.6 in storm-shaded-deps, 1.4 in external/storm-solr, 2.5 in storm-submit-tools and etc... Such version inconsistencies may cause unnecessary maintenance effort in the long run. For example, if two modules become inter-dependent, library version conflict may happen. It has already become a common issue and hinders development progress. Thus a version harmonization is necessary.<br/>
 <br/>
Provided we applied a version harmonization, I calculated the cost it may have to harmonize to all upper versions including an up-to-date one. The cost refers to POM config changes and API invocation changes. Take commons-io:commons-io for example, if we harmonize all the library versions into 2.6. The concern is, how much should the project code adapt to the newer library version. We list an effort table to quantify the harmonization cost.<br/>
 <br/>
The effort table is listed below. It shows the overall harmonization effort by modules. The columns represents the number of library APIs and API calls(NA,NAC), deleted APIs and API calls(NDA,NDAC) as well as modified API and API calls(NMA,NMAC). Modified APIs refers to those APIs whose call graph is not the same as previous version. Take the first row for example, if upgrading the library into version 2.6. Given that 9 APIs is used in module storm-server, 0 of them is deleted in a recommended version(which will throw a NoMethodFoundError unless re-compiling the project), 0 of them is regarded as modified which could break the former API contract.</p>

<div class='table-wrap'>
<table class='confluenceTable'><tbody>
<tr>
<th class='confluenceTh'>Index</th>
<th class='confluenceTh'>Module</th>
<th class='confluenceTh'>NA(NAC)</th>
<th class='confluenceTh'>NDA(NDAC)</th>
<th class='confluenceTh'>NMA(NMAC)</th>
</tr>
<tr>
<td class='confluenceTd'>1</td>
<td class='confluenceTd'>storm-server</td>
<td class='confluenceTd'>9(15)</td>
<td class='confluenceTd'>0(0)</td>
<td class='confluenceTd'>0(0)</td>
</tr>
<tr>
<td class='confluenceTd'>2</td>
<td class='confluenceTd'>integration-test</td>
<td class='confluenceTd'>4(4)</td>
<td class='confluenceTd'>0(0)</td>
<td class='confluenceTd'>0(0)</td>
</tr>
<tr>
<td class='confluenceTd'>3</td>
<td class='confluenceTd'>storm-submit-tools</td>
<td class='confluenceTd'>1(1)</td>
<td class='confluenceTd'>0(0)</td>
<td class='confluenceTd'>1(1)</td>
</tr>
<tr>
<td class='confluenceTd'>4</td>
<td class='confluenceTd'>..</td>
<td class='confluenceTd'>..</td>
<td class='confluenceTd'>..</td>
<td class='confluenceTd'>..</td>
</tr>
</tbody></table>
</div>

<p> <br/>
Also we provided another table to show the potential files that may be affected due to library API change, which could help to spot the concerned API usage and rerun the test cases. The table is listed below.</p>

<div class='table-wrap'>
<table class='confluenceTable'><tbody>
<tr>
<th class='confluenceTh'>Module</th>
<th class='confluenceTh'>File</th>
<th class='confluenceTh'>Type</th>
<th class='confluenceTh'>API</th>
</tr>
<tr>
<td class='confluenceTd'>storm-submit-tools</td>
<td class='confluenceTd'>storm-submit-tools/src/test/java/org/apache/storm/submit/dependency/DependencyResolverTest.java</td>
<td class='confluenceTd'>modify</td>
<td class='confluenceTd'>org.apache.commons.io.FileUtils.deleteQuietly(java.io.File)</td>
</tr>
</tbody></table>
</div>

<p> </p>

<p> <br/>
As for false consistency, take org.apache.commons commons-lang3 jar for example. The library is declared in version 3.3 in all modules. However they are declared differently. As components are developed in parallel, if one single library version is updated, which could become inconsistent as mentioned above, may cause above-mentioned inconsistency issues</p>

<p>If you are interested, you can have a more complete and detailed report in the attached PDF file.</p>

---
<details><summary><i>Originally reported by <a title="Kaifeng Huang" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=calvinhkf">calvinhkf</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-3593" target="_blank">Inconsistent library versions notice.</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2023-11-23T11:53:06+00:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rzo1">rzo1</a>:</i>
<p>Looks like this is outdated as we updated a lot of dependencies and fixed duplicates since this was reported. Feel free to open a new Jira.</p>
